File: fsharp/isbn-verifier/IsbnVerifier.fs
module IsbnVerifier

open System
open System.Text.RegularExpressions

let isValid (isbn: string) : bool = 
    Regex.IsMatch(isbn, @"^(\d-?){9}(\d|X)$") &&
    (isbn 
        |> Seq.filter(Char.IsLetterOrDigit) 
        |> Seq.mapi(fun i c -> ( if c = 'X' then 10 else (int) (Char.GetNumericValue(c)) ) * (10 - i) )
        |> Seq.sum
    ) % 11 = 0

File: csharp/isbn-verifier/IsbnVerifier.cs
using System;
using System.Linq;
using System.Text.RegularExpressions;
using System.Collections.Generic;

public static class IsbnVerifier
{
    static Regex VALID_ISBN = new Regex(@"^(\d-?){9}(\d|X)$");
    public static bool IsValid(string number)
    {
        if (!VALID_ISBN.IsMatch(number)) return false;
        return number
                .Where(c => char.IsDigit(c) || c == 'X')
                .Reverse()
                .Zip(Enumerable.Range(1, 10))
                .Aggregate(
                    0, (acc, tuple) => 
                       acc + ( tuple.First == 'X' ? 10 : (int) char.GetNumericValue(tuple.First) ) * tuple.Second
                    ) 
                % 11 == 0;
    }
}